tcms/issuetracker/tests/test_bitbucket.py

```diff
@@ -0,0 +1,185 @@
+from tcms.tests.factories import TestExecutionFactory
+
+
+@unittest.skipUnless(os.getenv('TEST_BUGTRACKER_INTEGRATION'),
+                     'Bug tracker integration testing not enabled')
+class TestBitBucketIntegration(APITestCase):
+    existing_bug_id = 1
+    existing_bug_url = 'https://bitbucket.org/kiwitcms/integration/issues/1'
+
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+        self.execution_1 = TestExecutionFactory()
+        self.execution_1.case.summary = "Tested at " + timezone.now().isoformat()
+        self.execution_1.case.text = "Given-When-Then"
+        self.execution_1.case.save()  # will generate history object
+        self.execution_1.run.summary = "Automated TR for BitBucket integration on " + \
+                                       timezone.now().isoformat()
+        self.execution_1.run.save()
+
+        self.component = ComponentFactory(name='BitBucket integration',
+                                          product=self.execution_1.run.plan.product)
+        self.execution_1.case.add_component(self.component)
+
+        bug_system = BugSystem.objects.create(  # nosec:B106:hardcoded_password_funcarg
+            name='BitBucket for kiwitcms/test-bitbucket-integration',
+            tracker_type='tcms.issuetracker.bitbucket.BitBucket',
+            base_url='https://bitbucket.org/kiwitcms/integration',
+            api_username=os.getenv('BITBUCKET_INTEGRATION_API_USERNAME'),
+            api_password=os.getenv('BITBUCKET_INTEGRATION_API_PASSWORD')
+        )
+        self.integration = BitBucket(bug_system, None)
+
+    def test_bug_id_from_url(self):
+        result = self.integration.bug_id_from_url(self.existing_bug_url)
+        self.assertEqual(self.existing_bug_id, result)
+
+    def test_details(self):
+        result = self.integration.details(self.existing_bug_url)
+
+        self.assertEqual('Hello World', result['title'])
+        self.assertIn(
+            "First public bug here",
+            result['description'])
+
+    def test_auto_update_bugtracker(self):
+        initial_comments = self.integration.rpc.get_comments(self.existing_bug_id)
+
+        # simulate user adding a new bug URL to a TE and clicking
+        # 'Automatically update bug tracker'
+        result = self.rpc_client.TestExecution.add_link({
+            'execution_id': self.execution_1.pk,
+            'is_defect': True,
+            'url': self.existing_bug_url,
+        }, True)
+
+        # making sure RPC above returned the same URL
+        self.assertEqual(self.existing_bug_url, result['url'])
+
+        # wait until comments have been refreshed b/c this seem to happen async
+        initial_comments_length = initial_comments['size']
+        current_comments_length = initial_comments_length
+        while current_comments_length != initial_comments_length + 1:
+            comments = self.integration.rpc.get_comments(self.existing_bug_id)
+            current_comments_length = comments['size']
```

Comment:
IMO it is good to remove the comments on tearDown or at the end of the test b/c they will accumulate over time and will cause subsequent tests to execute slower. That's what we do for GitHub & JIRA tests.